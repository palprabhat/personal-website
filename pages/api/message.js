const FormData = require("form-data");

const validateHuman = async (token) => {
  const secret = process.env.RECAPTCHA_SITE_SECRET;
  const response = await fetch(
    `https://www.google.com/recaptcha/api/siteverify?secret=${secret}&response=${token}`,
    {
      method: "POST",
    }
  );
  const data = await response.json();
  return data.success;
};

export default async function handler(req, res) {
  if (req.method === "POST") {
    const reqData = JSON.parse(req.body);

    const human = await validateHuman(reqData.reToken);
    console.log(human);
    if (!human) {
      res.status(400);
      res.json({ errors: ["Please, you're not fooling us, bot."] });
      return;
    }

    try {
      const formData = new FormData();

      Object.entries(reqData).map(([key, value]) => {
        formData.append(key, value);
      });

      const response = await fetch(process.env.CONTACT_API_URL, {
        method: "POST",
        body: formData,
      });
      const responseData = await response.json();

      if (responseData.result === "success") {
        res.status(200).send({ data: responseData.data });
        return;
      }

      res.status(500).send({ error: "Unable to send message" });
    } catch (err) {
      console.error(err);
      res.status(500).send({ error: "Unable to send message" });
    }
  } else {
    res.status(405).send({ error: "Request method not supported" });
  }
}
