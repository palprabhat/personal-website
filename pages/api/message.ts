/* eslint-disable @typescript-eslint/no-explicit-any */
import { NextApiRequest, NextApiResponse } from "next";
import FormData from "form-data";

type RequestBody = {
  name: string;
  email: string;
  subject: string;
  message: string;
  reToken: string;
};

type SendMessage = (reqBody: RequestBody) => Promise<string | null>;
type ValidateHuman = (token: string) => Promise<boolean>;

const cleanFromData = (formData: URLSearchParams) => {
  formData.forEach((_value, key) => formData.delete(key));
  formData.delete("_valueLength");
  formData.delete("writable");
  formData.delete("dataSize");
  formData.delete("pauseStreams");
  formData.delete("_streams");
  formData.delete("_insideLoop");
};

const validateHuman: ValidateHuman = async (token) => {
  const secret = process.env.RECAPTCHA_SITE_SECRET;
  const response = await fetch(
    `https://www.google.com/recaptcha/api/siteverify?secret=${secret}&response=${token}`,
    {
      method: "POST",
    }
  );
  const data = await response.json();
  return data.success as boolean;
};

const sendMessage: SendMessage = async (reqBody): Promise<any> => {
  try {
    const formData = new URLSearchParams(new FormData() as any);

    cleanFromData(formData);

    Object.entries(reqBody).map(([key, value]) => {
      if (key !== "reToken") formData.append(key, value);
    });

    const response = await fetch(process.env.CONTACT_API_URL, {
      method: "POST",
      body: formData,
    });

    const responseData = await response.json();

    if (responseData.result === "success") {
      return responseData.data;
    }
    return null;
  } catch (err) {
    console.error(err);
    return null;
  }
};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> {
  if (req.method === "POST") {
    const reqBody = JSON.parse(req.body) as RequestBody;

    const human = await validateHuman(reqBody.reToken);
    if (!human) {
      res.status(400).send({ error: "Could not verify you as human" });
      return;
    }

    const resData = await sendMessage(reqBody);

    if (resData) {
      res.status(200).send({ data: resData });
    } else {
      res.status(500).send({ error: "Unable to send message" });
    }
  } else {
    res.status(405).send({ error: "Request method not supported" });
  }
}
