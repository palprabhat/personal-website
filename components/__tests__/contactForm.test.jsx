import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ContactForm from "../contactForm";
import { mockFetch, mockMessage } from "../../pages/api/__mock__/message.mock";

const renderContactFormComponent = () => {
  const mockSubmitted = jest.fn();
  const { getByText, getByPlaceholderText, ...utils } = render(
    <ContactForm submitted={mockSubmitted} />
  );

  const name = getByPlaceholderText("Name");
  const email = getByPlaceholderText("Email");
  const subject = getByPlaceholderText("Subject");
  const message = getByPlaceholderText("Message");
  const submitButton = getByText("Send message");

  const fillForm = () => {
    userEvent.type(name, mockMessage.name);
    userEvent.type(email, mockMessage.email);
    userEvent.type(subject, mockMessage.subject);
    userEvent.type(message, mockMessage.message);
  };

  return {
    name,
    email,
    subject,
    message,
    submitButton,
    mockSubmitted,
    fillForm,
    ...utils,
  };
};

beforeAll(() => jest.spyOn(window, "fetch"));
beforeEach(() => window.fetch.mockImplementation(mockFetch));
afterEach(() => window.fetch.mockClear());

describe("ContactForm Component", () => {
  it("should render correct content", async () => {
    const {
      name,
      email,
      subject,
      message,
      submitButton,
      baseElement,
      queryAllByText,
    } = renderContactFormComponent();

    expect(name).toBeInTheDocument();
    expect(email).toBeInTheDocument();
    expect(subject).toBeInTheDocument();
    expect(message).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();

    await waitFor(() => {
      const errors = queryAllByText(/^Please /);
      expect(errors.length).toBe(0);
    });

    expect(baseElement).toMatchSnapshot();
  });

  it("should not submit an empty form", async () => {
    const {
      queryAllByText,
      submitButton,
      baseElement,
      mockSubmitted,
    } = renderContactFormComponent();

    userEvent.click(submitButton);

    await waitFor(() => {
      expect(mockSubmitted).toHaveBeenCalledTimes(0);

      const errors = queryAllByText(/^Please /);
      expect(errors.length).toBe(4);
    });

    expect(baseElement).toMatchSnapshot();
  });

  it("should not submit with invalid email", async () => {
    const {
      email,
      queryByText,
      submitButton,
      fillForm,
      mockSubmitted,
    } = renderContactFormComponent();

    fillForm();
    userEvent.type(email, "test@");
    userEvent.click(submitButton);

    await waitFor(() => {
      const errors = queryByText("Please provide a valid email");
      expect(errors).toBeInTheDocument();

      expect(mockSubmitted).toHaveBeenCalledTimes(0);
    });
  });

  it("should be able to fill form", () => {
    const {
      name,
      email,
      subject,
      message,
      fillForm,
      queryAllByText,
    } = renderContactFormComponent();

    fillForm();

    expect(name.value).toBe(mockMessage.name);
    expect(email.value).toBe(mockMessage.email);
    expect(subject.value).toBe(mockMessage.subject);
    expect(message.value).toBe(mockMessage.message);

    const errors = queryAllByText(/^Please /);
    expect(errors.length).toBe(0);
  });

  it("should submit with valid form data", async () => {
    const {
      submitButton,
      fillForm,
      mockSubmitted,
      baseElement,
    } = renderContactFormComponent();

    fillForm();

    userEvent.click(submitButton);
    await waitFor(() => {
      expect(mockSubmitted).toHaveBeenCalledTimes(1);
    });

    expect(baseElement).toMatchSnapshot();
  });

  it("should show error message on submission failed", async () => {
    window.fetch.mockImplementation((url) => mockFetch(url, { error: true }));

    const {
      submitButton,
      fillForm,
      mockSubmitted,
      queryByText,
      baseElement,
    } = renderContactFormComponent();

    fillForm();

    userEvent.click(submitButton);
    await waitFor(() => {
      const errors = queryByText("Something went wrong");
      expect(errors).toBeInTheDocument();

      expect(mockSubmitted).toHaveBeenCalledTimes(0);
    });

    expect(baseElement).toMatchSnapshot();
  });

  it("should handle unexpected exception", async () => {
    window.fetch.mockImplementation((url) =>
      mockFetch(url, { exception: true })
    );

    const {
      submitButton,
      fillForm,
      mockSubmitted,
      queryByText,
      baseElement,
    } = renderContactFormComponent();

    fillForm();

    userEvent.click(submitButton);
    await waitFor(() => {
      const errors = queryByText("Something went wrong! Please try again.");
      expect(errors).toBeInTheDocument();

      expect(mockSubmitted).toHaveBeenCalledTimes(0);
    });

    expect(baseElement).toMatchSnapshot();
  });
});
