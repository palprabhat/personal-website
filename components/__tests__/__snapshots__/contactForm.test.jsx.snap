// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm Component should handle unexpected exception 1`] = `
<body>
  <script
    src="https://www.google.com/recaptcha/api.js?onload=onloadcallback&render=explicit"
  />
  <div>
    <form
      class="w-full undefined"
    >
      <div
        class="my-2"
      >
        <input
          aria-label="Name"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="name"
          placeholder="Name"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <input
          aria-label="Email"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="email"
          placeholder="Email"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <input
          aria-label="Subject"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="subject"
          placeholder="Subject"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <textarea
          aria-label="Message"
          class="border-none rounded-3xl px-4 py-2 w-full outline-none resize-none"
          name="message"
          placeholder="Message"
          rows="3"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div />
      <button
        class="btn-primary"
        type="submit"
      >
        Send message
      </button>
      <div
        class="text-red-600 text-sm text-left pl-5 h-6"
      >
        Something went wrong! Please try again.
      </div>
    </form>
  </div>
</body>
`;

exports[`ContactForm Component should not submit an empty form 1`] = `
<body>
  <script
    src="https://www.google.com/recaptcha/api.js?onload=onloadcallback&render=explicit"
  />
  <div>
    <form
      class="w-full undefined"
    >
      <div
        class="my-2"
      >
        <input
          aria-label="Name"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="name"
          placeholder="Name"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        >
          Please provide your name
        </div>
      </div>
      <div
        class="my-2"
      >
        <input
          aria-label="Email"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="email"
          placeholder="Email"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        >
          Please provide your email
        </div>
      </div>
      <div
        class="my-2"
      >
        <input
          aria-label="Subject"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="subject"
          placeholder="Subject"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        >
          Please add a subject
        </div>
      </div>
      <div
        class="my-2"
      >
        <textarea
          aria-label="Message"
          class="border-none rounded-3xl px-4 py-2 w-full outline-none resize-none"
          name="message"
          placeholder="Message"
          rows="3"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        >
          Please add a message
        </div>
      </div>
      <div />
      <button
        class="btn-primary"
        type="submit"
      >
        Send message
      </button>
      
    </form>
  </div>
</body>
`;

exports[`ContactForm Component should render correct content 1`] = `
<body>
  <div>
    <form
      class="w-full undefined"
    >
      <div
        class="my-2"
      >
        <input
          aria-label="Name"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="name"
          placeholder="Name"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <input
          aria-label="Email"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="email"
          placeholder="Email"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <input
          aria-label="Subject"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="subject"
          placeholder="Subject"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <textarea
          aria-label="Message"
          class="border-none rounded-3xl px-4 py-2 w-full outline-none resize-none"
          name="message"
          placeholder="Message"
          rows="3"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div />
      <button
        class="btn-primary"
        type="submit"
      >
        Send message
      </button>
      
    </form>
  </div>
  <script
    src="https://www.google.com/recaptcha/api.js?onload=onloadcallback&render=explicit"
  />
</body>
`;

exports[`ContactForm Component should show error message on submission failed 1`] = `
<body>
  <script
    src="https://www.google.com/recaptcha/api.js?onload=onloadcallback&render=explicit"
  />
  <div>
    <form
      class="w-full undefined"
    >
      <div
        class="my-2"
      >
        <input
          aria-label="Name"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="name"
          placeholder="Name"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <input
          aria-label="Email"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="email"
          placeholder="Email"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <input
          aria-label="Subject"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="subject"
          placeholder="Subject"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <textarea
          aria-label="Message"
          class="border-none rounded-3xl px-4 py-2 w-full outline-none resize-none"
          name="message"
          placeholder="Message"
          rows="3"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div />
      <button
        class="btn-primary"
        type="submit"
      >
        Send message
      </button>
      <div
        class="text-red-600 text-sm text-left pl-5 h-6"
      >
        Something went wrong
      </div>
    </form>
  </div>
</body>
`;

exports[`ContactForm Component should submit with valid form data 1`] = `
<body>
  <script
    src="https://www.google.com/recaptcha/api.js?onload=onloadcallback&render=explicit"
  />
  <div>
    <form
      class="w-full undefined"
    >
      <div
        class="my-2"
      >
        <input
          aria-label="Name"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="name"
          placeholder="Name"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <input
          aria-label="Email"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="email"
          placeholder="Email"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <input
          aria-label="Subject"
          class="border-none rounded-3xl w-full px-5 py-2 outline-none"
          name="subject"
          placeholder="Subject"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div
        class="my-2"
      >
        <textarea
          aria-label="Message"
          class="border-none rounded-3xl px-4 py-2 w-full outline-none resize-none"
          name="message"
          placeholder="Message"
          rows="3"
        />
        <div
          class="text-red-600 text-sm text-left pl-5 h-6"
        />
      </div>
      <div />
      <button
        class="btn-primary"
        type="submit"
      >
        Send message
      </button>
      
    </form>
  </div>
</body>
`;
